unit_testing/LightController/LightController_tests.cpp

```diff
@@ -0,0 +1,432 @@
+#include "fake_USBprint.h"
+
+
+/* Real supporting units */
+#include "CAProtocol.c"
+#include "CAProtocolStm.c"
+
+/* UUT */
+#include "LightController.c"
+
+using ::testing::Contains;
+using ::testing::ElementsAre;
+using namespace std;
+
+/***************************************************************************************************
+** DEFINES
+***************************************************************************************************/
+
+/* Allow a range of single line container tests */
+#define EXPECT_READ_USB(x) { \
+    vector<string>* ss = hostUSBread(); \
+    EXPECT_THAT(*ss, (x)); \
+    delete ss; \
+}
+
+/* Allow a range of single line container tests */
+#define EXPECT_FLUSH_USB(x) { \
+    vector<string>* ss = hostUSBread(true); \
+    EXPECT_THAT(*ss, (x)); \
+    delete ss; \
+}
+
+
+/***************************************************************************************************
+** TEST FIXTURES
+***************************************************************************************************/
+
+class LightControllerTest: public ::testing::Test 
+{
+    protected:
+        /*******************************************************************************************
+        ** METHODS
+        *******************************************************************************************/
+        LightControllerTest()
+        {
+            /* OTP code to allow initialisation of the board to pass */
+            /* TODO: Make this use a define for the board release, so that tests fail if the someone
+            ** forgets to update the version numbers */
+            BoardInfo bi = {
+                .v2 = {
+                    .otpVersion = OTP_VERSION_2,
+                    .boardType  = LightController,
+                    .subBoardType = 0,
+                    .reserved = {0},
+                    .pcbVersion = {
+                        .major = 1,
+                        .minor = 1
+                    },
+                    .productionDate = 0
+                }
+            };
+            HAL_otpWrite(&bi);
+            forceTick(0);
+            hostUSBConnect();
+        }
+
+        void writeLightControllerMessage(const char * msg)
+        {
+            hostUSBprintf(msg);
+            LightControllerLoop(bootMsg);
+        }
+
+        void simTick(int numTicks = 1)
+        {
+            for(int i = tickCounter + 1; i <= tickCounter + numTicks; i++) 
+            {
+                HAL_TIM_PeriodElapsedCallback(&htim2);
+            }
+            tickCounter = tickCounter + numTicks;
+        }
+
+        void goToTick(int tickDest) {
+            while(tickCounter < tickDest) {
+                simTick();
+            }
+
+            // Reset counter
+            if (tickDest == 0){
+                while (tickCounter < TICK_COUNTER_OVERFLOW){
+                    simTick();
+                }
+                tickCounter = 0;
+            }
+        }
```

Comment:
We should move these tick functions (and the USB macros) to a unit testing util file. Maybe next time if you're in a hurry with this PR

Reply:
From discussion the USB macros have been moved into fake_USBprint.h, but the tick functions differ so they remain as they are.